File: mac/Marquee/API/Models/ImageRef.swift
import Foundation

extension API {
    /// An artwork path from the API (`posterPath`, `backdropPath`,
    /// `profilePath`, `logoPath`, `stillPath`). Usually a TMDb path
    /// (`"/abc.jpg"`), but a poster or backdrop can be a full `https://` URL
    /// (a TheTVDB fallback) — `url(_:)` handles both, like the website's
    /// `tmdbImageUrl` (api-v1.md deviation 1).
    struct ImageRef: Codable, Hashable, Sendable, CustomStringConvertible, ExpressibleByStringLiteral {
        /// TMDb's image widths. Posters: w92–w780; profiles: w45, w185, h632;
        /// backdrops: w300, w780, w1280; logos: w45–w500; stills: w92–w300.
        enum Size: String, Sendable, CaseIterable {
            case w45, w92, w154, w185, w300, w342, w500, w780, w1280, h632, original
        }

        /// The raw value as the server sent it.
        let path: String

        init(_ path: String) {
            self.path = path
        }

        init(stringLiteral value: String) {
            self.init(value)
        }

        init(from decoder: Decoder) throws {
            path = try decoder.singleValueContainer().decode(String.self)
        }

        func encode(to encoder: Encoder) throws {
            var container = encoder.singleValueContainer()
            try container.encode(path)
        }

        var description: String { path }

        /// Already a full URL; `size` doesn't apply.
        var isAbsolute: Bool {
            path.hasPrefix("https://") || path.hasPrefix("http://")
        }

        /// The image at `size`; nil for an empty path.
        func url(_ size: Size = .w500) -> URL? {
            Self.url(path, size: size)
        }

        /// The same rule for a plain string (or nil).
        static func url(_ path: String?, size: Size = .w500) -> URL? {
            guard let path, !path.isEmpty else { return nil }
            if path.hasPrefix("https://") || path.hasPrefix("http://") { return URL(string: path) }
            return URL(string: "https://image.tmdb.org/t/p/\(size.rawValue)\(path.hasPrefix("/") ? path : "/" + path)")
        }
    }
}

extension Optional where Wrapped == API.ImageRef {
    /// `card.posterPath.url(.w342)` without unwrapping first.
    func url(_ size: API.ImageRef.Size = .w500) -> URL? {
        self?.url(size)
    }
}

File: windows/Marquee.Core/Models/ImageRef.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Marquee.Core.Models;

/// <summary>
/// TMDb's image widths. Posters: w92 to w780; profiles: w45, w185, h632;
/// backdrops: w300, w780, w1280; logos: w45 to w500; stills: w92 to w300.
/// </summary>
public enum ImageSize
{
    W45,
    W92,
    W154,
    W185,
    W300,
    W342,
    W500,
    W780,
    W1280,
    H632,
    Original,
}

public static class ImageSizeExtensions
{
    /// <summary>The path segment TMDb expects: <c>w500</c>, <c>original</c>.</summary>
    public static string Segment(this ImageSize size) => size.ToString().ToLowerInvariant();
}

/// <summary>
/// An artwork path from the API (<c>posterPath</c>, <c>backdropPath</c>,
/// <c>profilePath</c>, <c>logoPath</c>, <c>stillPath</c>). Usually a TMDb
/// path (<c>"/abc.jpg"</c>), but a poster or backdrop can be a full
/// <c>https://</c> URL (a TheTVDB fallback). <see cref="Url"/> handles both,
/// like the website's <c>tmdbImageUrl</c> (api-v1.md deviation 1).
/// </summary>
[JsonConverter(typeof(ImageRefConverter))]
public readonly record struct ImageRef(string Path)
{
    public const string TmdbBase = "https://image.tmdb.org/t/p/";

    /// <summary>Already a full URL; a size doesn't apply.</summary>
    public bool IsAbsolute => IsAbsolutePath(Path);

    /// <summary>The image at <paramref name="size"/>; null for an empty path.</summary>
    public Uri? Url(ImageSize size = ImageSize.W500) => UrlFor(Path, size);

    /// <summary>The same rule for a plain string (or null).</summary>
    public static Uri? UrlFor(string? path, ImageSize size = ImageSize.W500)
    {
        if (string.IsNullOrEmpty(path))
        {
            return null;
        }
        if (IsAbsolutePath(path))
        {
            return Uri.TryCreate(path, UriKind.Absolute, out var absolute) ? absolute : null;
        }
        var relative = path.StartsWith('/') ? path : "/" + path;
        return Uri.TryCreate(TmdbBase + size.Segment() + relative, UriKind.Absolute, out var built) ? built : null;
    }

    public override string ToString() => Path;

    public static implicit operator ImageRef(string path) => new(path);

    private static bool IsAbsolutePath(string path) =>
        path.StartsWith("https://", StringComparison.Ordinal) || path.StartsWith("http://", StringComparison.Ordinal);
}

public static class ImageRefExtensions
{
    /// <summary><c>card.PosterPath.Url(ImageSize.W342)</c> without unwrapping first.</summary>
    public static Uri? Url(this ImageRef? image, ImageSize size = ImageSize.W500) => image?.Url(size);
}

/// <summary>An <see cref="ImageRef"/> is a bare string on the wire.</summary>
public sealed class ImageRefConverter : JsonConverter<ImageRef>
{
    public override ImageRef Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        if (reader.TokenType != JsonTokenType.String)
        {
            throw new JsonException($"Expected an image path string, got {reader.TokenType}.");
        }
        return new ImageRef(reader.GetString()!);
    }

    public override void Write(Utf8JsonWriter writer, ImageRef value, JsonSerializerOptions options) =>
        writer.WriteStringValue(value.Path);
}
